ls1intum/Artemis
# Improve performance of semi-automatic modeling assessments

### Checklist

### Motivation and Context


The new automatic assessment mechanism implemented in https://github.com/ls1intum/Artemis/pull/548 has some major performance issues when there are a lot of modeling submissions without assessment. Especially, when a manual assessment gets submitted by a tutor, it has a high impact on users. The automatic assessment of all submissions is triggered after a manual assessment is submitted which results in a response time of ~25-30 seconds after a submit (for 1000 existing submissions in the database).

### Description

To cope with this problem, we change the storage of the automatic results. After an automatic result is created, it will not get stored to the database anymore. Instead, a ConcurrentHashMap will be used in Compass to store the automatic results. As soon as a submission is locked for assessment, the corresponding automatic result will be retrieved from the hash map and stored to the database. Additionally, we adjust the process of calculating the next optimal submissions that should be assessed next. 

The changes in detail:
- Store automatic results in hash map instead of database to reduce the number of database calls.
- Retrieve automatic result from hash map and store it together with submission to database when a submission is locked for assessment.
- Remove automatic result from hash map after it is stored to the database (and after an assessment is canceled).
- Load submission together with result, feedback, assessor, participation and participation results in one database access to reduce the number of database calls.
- Load all submissions without assessment from the database at once when the automatic assessment of all submissions is triggered (e.g. after a manual assessment submit) instead of loading each submission in an a separate iteration.
- Optimal submissions: When the number of cached optimal submissions is smaller than a configured threshold of 10 submissions, 10 new optimal submissions are calculated by the `ModelSelector` and loaded into the cache, instead of just loading one new optimal submission every time. By that, the user has to wait for the calculation only every 10th assessment on average.
- Optimal submissions: Every time 10 new optimal submissions are calculated by the `ModelSelector`, they are calculated at once, instead of doing the calculation 10 times in a row.
- Optimal submissions: If Compass could assign automatic feedback to all the elements of each submission, 10 random submissions are now chosen, as there are no optimal submissions anymore if everything is already assessed automatically. Also, there was a small "bug" in this case that lead to the situation that every submission without an assessment was compared to every other submission without assessment resulting in tremendous calculation effort. Selecting a new optimal submission took about 12 seconds for ~1000 existing submissions. By fixing this we reduced the calculation time to ~1 second.

Through these changes the following performance improvements could be achieved:
- The response time after submitting an assessment decreased from 25-30 seconds to ~1.5 seconds.
- The time of initializing Compass (after a server restart) decreased from over 5 minutes to ~4.5 second.
- The time of loading new submissions into the next optimal submission cache takes ~1 second, now 10 submissions are loaded at once instead of only one.

### Steps for Testing

Precondition: a high number of submissions (~1000) that are not assessed exist in the database for one exercise.

1. Log in to ArTEMiS as a tutor
2. Navigate to the tutor exercise dashboard of the corresponding exercise
3. Submit an assessment for a submission
4. Check response time
